docs/yul.rst

```diff
@@ -952,6 +952,26 @@ option.
 
 See :ref:`Using the Commandline Compiler <commandline-compiler>` for details about the Solidity linker.
 
+memoryguard
+^^^^^^^^^^^
+
+This function is available in the EVM dialect with objects. The caller of
+``let ptr := memoryguard(size)`` (where ``size`` has to be a literal number)
+promises that they only use memory in either the range ``[0, size)`` or the
+unbounded range above ``ptr``.
+
+Since the presence of a ``memoryguard`` call indicates that all memory access
+adheres to this restriction, it allows the optimizer to perform additional
+optimization steps, for example the stack limit evader, which attempts to move
+stack variables that would otherwise be unreachable to memory.
+
+The Yul optimizer promises to only use the memory range ``[size, ptr)`` for its purposes.
```

Comment:
Should we mention that it will modify the literal argument of `memoryguard` such that the resulting code again fulfisl the promise?

Reply:
I thought about it, but decided against it, because I thought it would probably mainly be confusing and in the end it's an implementation question that doesn't have user-facing consequences. But if you think it should be mentioned here, I'll do that.

Reply:
I think it is useful to understand the output of the optimizer.

Reply:
You can say something like in the current implementation it is done like that.